### ISSUE TITLE
Plan for Entity Framework Core 7.0

### ISSUE THREAD

**Issue by ajcvickers** [dotnet/efcore#26994]
Today we are excited to share with you [the plan for Entity Framework Core 7.0](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan). This issue contains a quick summary of the plan and acts as a place for you to leave feedback.

This plan brings together input from many stakeholders and outlines where and how we intend to invest in Entity Framework Core 7.0 (EF Core 7.0.) For brevity, EF Core 7.0 is also referred to as just EF7.

> **IMPORTANT**
> This plan is not a commitment; it will evolve as we continue to learn throughout the release. Some things not currently planned for EF7 may get pulled in. Some things currently planned for EF7 may get punted out.

# General information

EF Core 7.0 is the next release after EF Core 6.0 and is currently scheduled for release in November 2022 at the same time as .NET 7. There are no plans for an EF Core 6.1 release.

EF7 will align with the [.NET support policy](https://dotnet.microsoft.com/platform/support/policy/dotnet-core) and will therefore **not** be a long-term support (LTS) release.

EF7 currently targets .NET 6. This may be updated to .NET 7 as we near the release. EF7 does not target any .NET Standard version; for more information see [the future of .NET Standard](https://devblogs.microsoft.com/dotnet/the-future-of-net-standard/). EF7 will not run on .NET Framework.

## Themes

The large investments in EF7 will fall mainly under the following themes.

### Highly requested features

As always, a major input into the [planning process](https://docs.microsoft.com/ef/core/what-is-new/release-planning) comes from [votes (👍) for features on GitHub](https://github.com/dotnet/efcore/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

- [JSON columns](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#json-columns): Save and query into JSON-based documents stored in relational database columns.
- [Bulk updates](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#bulk-updates): Efficient, predicate-based updates for many database rows without loading data into memory.
- [Lifecycle hooks](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#lifecycle-hooks): Allow applications to react when interesting things happen in EF code.
- [Table-per-concrete-type (TPC) mapping](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#table-per-concrete-type-tpc-mapping): Map entities in a hierarchy to separate tables without taking the performance hit of TPT mapping.
- [Map CUD operations to stored procedures](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#map-cud-operations-to-stored-procedures): Use stored procedures to manage data modifications.
- [Value objects](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#value-objects): Applications can use DDD-style value objects in EF models.
- [Support value generation when using value converters](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#support-value-generation-when-using-value-converters): DDD-style encapsulated key types can make full use of automatically generated key values.
- [Raw SQL queries for unmapped types](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#raw-sql-queries-for-unmapped-types): Applications can execute more types of raw SQL query without dropping down to ADO.NET or using third-party libraries.
- [Database scaffolding templates](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#database-scaffolding-templates): The code generated by `dotnet ef database scaffold` can be fully customized.

### .NET platforms and ecosystem

Much of the work planned for EF7 involves improving the data access experience for .NET across different platforms and domains. This involves work in EF Core where needed, but also work in other areas to ensure a great experience across .NET technologies.

- [Distributed transactions](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#distributed-transactions): .NET Framework applications using distributed transactions can be ported to .NET 7.
- [EF Core tooling](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#ef-core-tooling): Ensure `dotnet ef` commands are easy to use and work with modern platforms and technologies.
- [EF Core and graphical user interfaces](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#ef-core-and-graphical-user-interfaces): Make it easy to build data-bound graphical applications with EF Core.
- [SqlServer.Core (Woodstar)](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#sqlservercore-woodstar): Fast, fully managed access to SQL Server and Azure SQL for modern .NET applications.
- [Azure Cosmos DB provider](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#azure-cosmos-db-provider): Continue to make EF Core the easiest and most productive way to work with [Azure Cosmos DB](https://azure.microsoft.com/services/cosmos-db/#overview).
- [Migrations experience](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#migrations-experience): Make it easy to get started with migrations and later use them effectively in CI/CD pipelines.
- [Trimming](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#trimming): Smaller applications that can be efficiently AOT compiled.
- [Evolve System.Linq.Expression](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#evolve-systemlinqexpression): Use modern C# language features in LINQ queries.
- [Translate new LINQ operators](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#translate-new-linq-operators): Use new LINQ operators when translating LINQ queries to SQL.
- [Open telemetry for ADO.NET providers](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#open-telemetry-for-ado-net-providers): Cross-platform, industry-standard telemetry that can be monitored in your tool of choice.
- [Enhancements to System.Data](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#enhancements-to-systemdata): Better low-level data access to benefit all higher-level code.
- [Research data access for cloud-native](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#research-data-access-for-cloud-native): Future evolution of .NET data access that supports modern approaches such as microservices and cloud native.

### Clear path forward from EF6

EF Core has always [supported many scenarios not covered by the legacy EF6 stack](https://docs.microsoft.com/ef/efcore-and-ef6/), as well as being generally much higher performing. However, EF6 has likewise supported [scenarios not covered by EF Core](https://docs.microsoft.com/ef/efcore-and-ef6/). EF7 will add support for many of these scenarios, allowing more applications to port from legacy EF6 to EF7. At the same time, we are planning a [comprehensive porting guide](https://github.com/dotnet/EntityFramework.Docs/issues/1180) for applications moving from legacy EF6 to EF Core.

## Performance

Great performance is a fundamental tenet of EF Core, lower-level data access, and indeed all of .NET. Every release includes significant work on improving performance.

- [Performance of database inserts and updates](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#performance-of-database-inserts-and-updates): High performance database inserts and updates from EF Core
- [TechEmpower composite score](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan#techempower-composite-score): High performing low-level data updates for all .NET applications.

# Find out more and give feedback

This post is a brief summary of the [full EF7 plan](https://docs.microsoft.com/ef/core/what-is-new/ef-core-7.0/plan). Please see the full plan for more information.

Your feedback on planning is important. The best way to indicate the importance of an issue is to vote (👍) for that [issue on GitHub](https://github.com/dotnet/efcore/issues). This data will then feed into the [planning process](https://docs.microsoft.com/ef/core/what-is-new/release-planning) for the next release.

In addition, please **comment on this issue** if you believe we are missing something that is critical for EF7, or are focusing on the wrong areas.

**Comment by benstevens48** [comment]
How about https://github.com/dotnet/efcore/issues/1387 (including/excluding specific properties when loading an entity from the database)? Together with bulk updates/deletes that's the main thing that's missing from EF Core for me.

**Comment by sveinungf** [comment]
Looks great! Have you also considered looking at #13617 (parameterizing `IN(...)` queries)? It would be helpful for performance on many of our workloads.

**Comment by atrauzzi** [comment]
This is very exciting, especially if the JSON column support allows for polymorphic strongly typed JSON,(via a discriminator field)  to come back. Think unions.

Also consider ways of supporting all the fantastic operators postgres has for JSON that can take advantage of indexes!

I'd like to offer a screen share if it would help the EF team to see what I'm doing currently.

**Comment by ErikEJ** [comment]
@sveinungf there are numerous ways to fix that

**Comment by davidroth** [comment]
Great plan! I was hoping to see [https://github.com/dotnet/efcore/issues/9630](https://github.com/dotnet/efcore/issues/9630) included though 😥

**Comment by HaroldH76** [comment]
Please, please, pretty please fix the async issues with varchar(max) columns, see for more info:
https://github.com/dotnet/efcore/issues/18221

If performance is important for EF Core then this one should really be fixed ;-)

And preferably fix it in EF core 6.

**Comment by roji** [comment]
@HaroldH76 the problem isn't in EF Core, it's in SQLClient, there's nothing really we can do to fix that...

**Comment by HaroldH76** [comment]
But both are Microsoft? So maybe EF Core team can push SqlClient team?
Or EF Core team can provide a workaround? I have seen suggestions about a custom DBInterceptor and caching some primary keys. Maybe EF Core Team can provide a setting for that instead that we have to write this "hack"?

**Comment by roji** [comment]
@HaroldH76 I'm not too sure what primary key caching you're referring to... the way to avoid this issue is to use sync APIs instead of async when dealing with large columns. We're in touch with the SqlClient people, but ultimately this has to be fixed there.

**Comment by ajcvickers** [comment]
@sveinungf Parametrization for IN queries is top of the list of bigger features that didn't make the cut.

**Comment by ajcvickers** [comment]
@davidroth Two questions:

* Would they have to be polymorphic?
* For owned references, would they have to be mapped to the same table as the owner?

**Comment by Martinn2** [comment]
Bulk operations would be great, but why only update? what about Bulk insert ? For my apps Bulk insert is more important.

**Comment by roji** [comment]
@Martinn2 EF Core already does a form of bulk insert - if you add multiple instances and then call SaveChanges, all insertions will happen in a single roundtrip, which is quite efficient.

More importantly, bulk update and delete are very similar in that the user provides a LINQ predicate defining which entities are to be updated/deleted, and that gets executed in the database (e.g. via SQL UPDATE or DELETE). Bulk insert is a completely different feature which doesn't have anything to do with that.

**Comment by Martinn2** [comment]
@roji Entity framework extensions, there is an online test in .NET Fiddle: https://dotnetfiddle.net/ttbri7

When I run this with 4000 entries, I got the following results:

![image](https://user-images.githubusercontent.com/45663203/146632044-89e20466-a60b-4d86-84ef-66bcf0d10e9c.png)

As you can see, SaveChanges is more than 6x slower. I understand that is is different from updating and deleting, but IMHO bulk insert is even more important

**Comment by roji** [comment]
@Martinn2 see #15059, and please let's keep any further conversation about bulk insert in that issue rather than here. tl;dr we do agree that EF Core's bulk insert could be optimized (see for example #9118 which is planned for 7.0), but our main focus here is database-only bulk update and delete, which are currently unsupported in EF Core and can make a huge difference.

**Comment by Meister1977** [comment]
Do you plan MySQL low level support without any additional nuget package?

**Comment by roji** [comment]
@Meister1977 what are you referring to exactly? There are already two EF Core providers for MySQL, see the [providers page](https://docs.microsoft.com/en-us/ef/core/providers/?tabs=dotnet-core-cli). All providers come as separate nuget packages, including the SQL Server and SQLite ones, which are maintained and delivered by the EF team.

**Comment by davidroth** [comment]
> @davidroth Two questions:
Would they have to be polymorphic?
For owned references, would they have to be mapped to the same table as the owner?

I am mostly interested in the ownership of polymporphic collections.
I described a canonical example in the following comment: https://github.com/dotnet/efcore/issues/9630#issuecomment-575174969

Regarding table mapping: In the OrderPosition example I would like to see TPH or TPT/C for the Position types:

- Order: Table "Orders"
- StandardPosition: Table "Positions" (Discriminator "standard")
- FancyPosition: Table "Positions" (Discriminator "fancy")

**Comment by cjblomqvist** [comment]
Just a few quick questions:

Is there any priority amount the different things planned? Are there some which are more "safe" for the 7 release and some which are less safe?

What's the reasoning behind not having a 6.1 release?

Thanks for your work!

**Comment by ajcvickers** [comment]
@cjblomqvist 

> Is there any priority amount the different things planned? Are there some which are more "safe" for the 7 release and some which are less safe?

That's a good question. I'll discuss with the team whether there is some way we can provide any additional information here. /cc @JeremyLikness 

> What's the reasoning behind not having a 6.1 release?

We align with .NET releases, which have converged to a regular schedule of one release per year. There are a lot of good reasons around this in general, but this isn't the best place to discuss them and I'm not the right person to express those reasons. The reason for explicitly stating this in the EF Core plan is to avoid people having any expectations to the contrary.

**Comment by flatproject** [comment]
A bit  😞  that nothing from #21459 has made it to EF7.  Need more votes...

**Comment by Kalle4242** [comment]
As well a bit 😞 that #4823 did not make it to EF7.

**Comment by VenkateshSrini** [comment]
Will we get any support for Event Sourcing like Marten (https://martendb.io/)

**Comment by VenkateshSrini** [comment]
How about the Distributed transaction for non windows platform?

**Comment by roji** [comment]
@VenkateshSrini event sourcing is something we already discussed and answered in #20306.

As for distributed transactions, [as the 7.0 plan says](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/plan#distributed-transactions), this is not something we plan to tackle at this point. The priority is currently to unblock people who cannot port their applications from .NET Framework because of the use of distributed transactions - and that's a Windows-only thing. If we see demand for a cross-platform version, we may tackle that at a later point.

**Comment by VenkateshSrini** [comment]
@roji ,
The issue [#20306](https://github.com/dotnet/efcore/issues/20306) was raised by me. The use case that I'm asking is, since like Postgres Json value storage is now enabled, based on the attribute value of JSON, I would like to perform aggregate on Event storage to get consolidated pictures. The event storage can still leverage the JSON storage to store the events. Good example is Marten with Postgres

**Comment by roji** [comment]
@VenkateshSrini I suggest opening an issue on the EFCore.PG repo with a very precise description of what you're looking for, including the SQL you'd like to see generated ("event sourcing" can be a pretty vague term). Note that JSON support is one of the main items in the EF Core 7.0 roadmap, though what actually will be in scope is still open, especially considering different JSON support across databases.

**Comment by brasky** [comment]
Hello team! I just watched the EF7 plan community standup, love all the work you do and love EF Core!

My team relies on the Cosmos DB Provider and is currently using it in conjunction with access keys as managed identities (i.e. DefaultAzureCredential()) are not supported. I just wanted to confirm that #26491 was still in the EF7 plan since internally there is a strong push to move away from Cosmos access keys.

Thanks!

**Comment by atrauzzi** [comment]
@ajcvickers @roji - Any chance https://github.com/dotnet/efcore/issues/4526 could sneak into the release?

**Comment by roji** [comment]
@atrauzzi #4526 is a "stretch goal" (that's the "consider-for-next-release" label) - that means that if we get through everything we plan and have time, it's a priority feature for getting into 7.0. But note that this is a pretty big feature with some serious design questions, since different databases have different upsert-like mechanisms - so not exactly a could candidate for "sneaking in" :)

**Comment by 3GDXC** [comment]
Q: will the issue around connection interceptors and the order of connection string validation be corrected in efcore 7.0 ?

**Comment by ajcvickers** [comment]
@3GDXC Issue #23085 is in the 7.0 milestone, which means it is currently planned for 7.0.

**Comment by sam-wheat** [comment]
@ajcvickers 

I would certainly appreciate it if you guys could find time to for this issue:
https://github.com/dotnet/efcore/issues/7776

I suppose it is a tough one because it's been open so long.  Fixing it would solve a ton of problems for me and hopefully many others

**Comment by ajcvickers** [comment]
@sam-wheat That issue only has 11 votes (👍), which means there are 142 open issues with more votes. This doesn't make it a high priority.

**Comment by AraHaan** [comment]
I have successfully migrated ``dotnet ef`` to an .NET 6 tool minimum (including it's internal ``ef.dll`` which is used by the PM console and by powershell). However I will need 1 more change to it and that is to do ``Major`` runtime roll-forwards on it as well (so if .NET 7.0.x runtime is installed, it will depend on 7.0.x instead of 6.0.x. 👌

**Comment by CsPauloNoob** [comment]
Windows Forms .NET CORE support pls

**Comment by roji** [comment]
@CsPauloNoob Windows Forms is supported by .NET Core, and works well with EF Core.

**Comment by rcbellamy** [comment]
Honestly, for the remainder of the EF7 timeframe and the entirety of the EF8 timeframe, Microsoft's only focus should be improving the documentation for the currently existing features.  Any Microsoft-employed developer responsible for developing EF Core that is not qualified to author documentation in proper English (and some of the current documentation is from authors that are not so qualified) should either be focused on bug fixes or improving the performance of the existing API, or should be **laid off**.  I cannot imagine what kind of stakeholders have any **legitimate** interest in further development of the API without better documentation of the already-existing API.

Furthermore, I am quite certain the God would not look kindly on any stakeholder allowing other "priorities" (for the EF Core team, that is) to get in the way of better documentation, so if Microsoft does encounter some stakeholder that it cares about whose representative is claiming that the EF Core developers should have "higher" priorities, I would strongly suggest that Microsoft speak to people at said stakeholder above that representative's head about how important those other priorities really are.

**Comment by AraHaan** [comment]
While documentations are great and all, anyone outside of Microsoft can fix them too (I think and hope this is the case) just like how anyone could bugfix efcore without problems if they know where a bug might be (report it first though so it can have a tracking issue and higher chance at getting merged).

As such I agree with Microsoft having a priority on adding new apis / changing older apis to suit the now as the now is different from the then in what is required to be done by efcore.

**Comment by rcbellamy** [comment]
@AraHaan , to fix something one must understand the problem and the solution, and that requires having the documentation.  You are ignoring the chicken-and-the-egg problem.  Plus, the documentation is Microsoft's job.

Adding/expanding APIs faster than they can/do maintain the documentation is not adding value, it is detracting from it.  It takes us from a state with more valid documentation to a state with less valid documentation.  **It is therefore Microsoft sabotaging its own products**, and the people responsible for it will spend time in hell for doing so.

**Comment by AraHaan** [comment]
While you might be right for saying that, it's not worth documenting preview features that may or may not make it into the final product until they release it (and then document it afterwards).

**Comment by rcbellamy** [comment]
@AraHaan You might have a point that preview features do not need to be fully documented, _yet_.  However, it is still very important that Microsoft include full documentation in the same budget that the feature itself is in.  Right now, much of the documentation is actually far behind fully-released updates to EF Core, and quite frankly, a lot of it has never been adequate regardless of the version of EF Core to which we compare the documentation.  Microsoft has authored some documents that they claim provide "overview[s]" without ever even claiming to provide a complete reference on those topics.  We need the complete reference in order to be able to figure out on our own how to handle situations that deviate from Microsoft's assumptions about how its products will be used 90% of the time.

Microsoft does not have any 90% customer.  Any software program being developed with commercial grade complexity is going to end up having some aspect of it that is not the same as how .Net (or any aspect thereof) is used 90% of the time, even if 95% of the code in the program-being-developed is just run-of-the-mill routine.  If all we as software developers were doing were things that were so simple that they don't ever deviate from the most common usage scenario, than we would all be out of a job because we as software developers would not be adding any value.  The notion that Microsoft has done its job so long as it has enabled us to understand 'the most common' way to use any given aspect of .Net is absurd, and yet it seems to be Microsoft's current mindset.

At least EF Core has documentation from the .Net Core/EF Core era.  I don't even want an ORM for the project I'm working on right now, but when I tried reading the ADO.Net documentation, what I found from Microsoft hadn't been updated to reflect the existence of .Net Core or other Core-era technologies such as EF Core.

**Comment by AraHaan** [comment]
Interesting how it was not updated to ADO.NET Core (which is Microsoft.Data.SqlClient).

**Comment by richlander** [comment]
Please keep your commentary civil @rcbellamy. Your choice of words doesn't align with our [code of conduct](https://github.com/dotnet/efcore/blob/main/.github/CODE_OF_CONDUCT.md). I hid your comment.

**Comment by rcbellamy** [comment]
@richlander Please identify a particular provision in the code of conduct.

**Comment by AraHaan** [comment]
![25AA14F0-D234-40E9-BF5D-68A5543F2BFB](https://user-images.githubusercontent.com/15173749/172254516-0a7febbe-f000-43d2-bf19-e747b0e5e38e.jpeg)
Take a look at these very closely.

**Comment by rcbellamy** [comment]
I disagree, and I note that the person that acted in a moderator capacity has not defended his determination with any citation to any particular provision.

**Comment by rcbellamy** [comment]
Perhaps you misunderstood what I was saying.  I was merely observing that the value of an ecosystem as a whole is a direct constraint on the ability of any member of that ecosystem to themselves produce value, and that as a result, pursuits that destroy the value of an ecosystem of which one is a member are shortsighted.  I understand that more senior members of this ecosystem often want some barrier to entry in order to ensure that their seniority corelates to an ability to provide more value than others provide.  However, the fact is that people retire, and even those that do not (as I understand science) do still eventually cease to be current members of the ecosystem (not an attack, just my understanding of science).  If new programmers are unable to enter the ecosystem, then the ecosystem ceases to be viable.  The lack of documentation from Microsoft goes way beyond preserving someone’s seniority for what is left of their natural career.  The barrier is so high that it is _**a direct threat to the ability of new programmers to enter the ecosystem, which is a direct threat to the ecosystem as a whole**_.  People that intend for the ecosystem to continue to be viable, yet would sidetrack Microsoft from providing minimally sufficient documentation, are acting in a shortsighted manner.

I am merely explaining why I see a particular decision as being shortsighted.  To call that “insulting” would make any difference of professional opinion qualify as an “insult[.]”  I have not so much as referred to any particular party, much less attacked their reasons for wanting to prioritize things in a way that I see as shortsighted.  I have only provided speculation for one motive, which I am not attacking.  I understand that position, and I do not intend to convey anything derogatory or insulting about such persons.  I am suggesting that those persons are intelligent, attempting to act in a rational manner, and perhaps almost perfectly doing so.  I am only suggesting that there might be a small miscalculation in their logic that ends up changing the result of their analysis.  Pointing out the possibility that one might have made a mistake is not an insult and is not derogatory.  By attempting to point it out, my actions demonstrate that I think highly enough of such persons for it to be worth my time to have this dialogue.

I take the time to say all of this because I do think that those persons are doing more harm than good for themselves without realizing it.  I am not at all suggesting that they do not have very legitimate reasons to want Microsoft to focus its investments elsewhere.  It just seems to me that people are overlooking the fact that Microsoft’s lack of documentation is a _**direct threat to the viability of the ecosystem as a whole**_.  If people consider that and disagree, that’s their call.  However, if they have simply been thinking of it as being a smaller problem without considering the point-of-view of the future of an ecosystem that may or may not be theirs (in that, without new programmers in the .Net ecosystem, the future ecosystem would not end up being .Net), then they do need to realize that their priorities have broader implications than the aspect of the ecosystem in which they work, and in which they are prioritizing one matter over another.

**Comment by roji** [comment]
@rcbellamy here's my personal take here, as a member of the EF team.

First, the parts in your comments which I find inappropriate in the professional setting, contrary to our code of conduct and  offensive are:

* "[...] the people responsible for it will spend time in hell for doing so." ([here](https://github.com/dotnet/efcore/issues/26994#issuecomment-1146434110))
* To a lesser extent: "I am quite certain the God would not look kindly on any stakeholder allowing other "priorities" [...] ([here](https://github.com/dotnet/efcore/issues/26994#issuecomment-1140351433))

Moving beyond these to your specific arguments. I've reread what you wrote, and to my understanding your complaint is about the low quailty of our documentation, and about the fact that we continue improving the product (with enhancements), while the documentation gap remains unaddressed ("Honestly, for the remainder of the EF7 timeframe and the entirety of the EF8 timeframe, Microsoft's only focus should be improving the documentation for the currently existing features."). I don't view these arguments as insulting or a violation of our CoC (as opposed to the two comments I cited above), but do think they are without merit and do not correspond to the reality of the vast majority of our users.

First, we invest a lot of effort into the EF Core documentation, and generally receive very good feedback about them from users; I don't think it's accurate that "much of the documentation is actually far behind fully-released updates to EF Core". In general, documentation is never perfect and never reaches 100% coverage, just like a product is never 100% bug-free. Just like with features and bugs, we prioritize writing documentation based on how many users are affected (and how severely), basing our estimation on user feedback (among other things). Important features are documented promptly (at the end of the release), while exotic edge cases may take a long time to get prioritized. I do not believe it's in our users' best interests to halt all feature work and write docs for scenarios which affect only a handful of users, and all the feedback we're getting from our users confirms this.

It's also important to understand that documentation isn't everything. In some cases, users aren't able to accomplish what they want exclusively by reading official documentation or a book, and get their answers on Stackoverflow, external blog posts, or by posting an issue on our github repo. I don't believe this should be viewed as a necessary failure, or something that needs to be fixed: in modern software development docs are one resource alongside several others. Of course, when we see that many users are reaching to us with the same problem, we take that as a signal that the docs should be improved, and do that.

Finally, rather than this general, abstract discussion, I'd recommend opening specific issues on concrete documentation gaps you find, on our [documentation repo](https://github.com/dotnet/EntityFramework.Docs). The only actual, concrete complaint I can see from you is in #28047 about the use of DataDictionary. In that issue, we pointed to docs touching on the general area, and though I agree those docs could be improved, yours is the only case AFAIK where a user tried to use DataDictionary in this way; as such, like I explained above, prioritizing those docs wouldn't be the interest of our users.

I do encourage you to open other issues on difficulties you encounter while using EF Core, and documentation gaps you identify. We'll be here to help you out, and where it makes sense, we'll also update our docs to help other users not encounter the same difficulties.

**Comment by rcbellamy** [comment]
You are correct that I am not providing very specific, detailed criticism of specific documentation failures.  However, I read about 400 pages of EF Core documentation, and if I were to submit specific criticisms, they would themselves fill an entire book.

You say that "[you] don't believe [that documentation that isn't fully complete] should [necessarily] be viewed as a []failure[.]"  I do not disagree with that.  However, when the development is so far ahead of the documentation that the gap between the documentation and the current API makes it so much more difficult to use the product that this difficulty has a larger, negative effect on the value of the product than the value of the new features, that is destroying your own product in the name of "expanding functionality."

You must realize that there are a lot of mistakes that would be readable by someone that already understands the material covered better than they understand the English language.  Microsoft needs to focus on finding people that actually do speak English fluently to explain their products to people that do not already know the products well enough to not need the documentation in the first place.  Technical experts that still fail to understand which ending is the right ending for the root word that they are using are not qualified to assess the quality of English-language documentation.

**Comment by AraHaan** [comment]
I would have to disagree on the fact that there would be no need for documentations, there are people like me who understand things better with samples or seeing it done from somebody first. It all depends on how people learn how to use frameworks properly and people always learn in different ways and that is why documentations exist.

**Comment by ilazarevski** [comment]
@ajcvickers I would like to check if an initial implementation is planned for **Azure Synapse** per [https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/plan#theme-net-platforms-and-ecosystem]?

**Comment by ErikEJ** [comment]
@ilazarevski Have you tried it with Azure Synapse - and what is not working?

**Comment by ilazarevski** [comment]
I didn’t test it, but first I wanted to check if it is already included in the preview. I will check and inform you about the results.

**Comment by roji** [comment]
@ilazarevski in principle, at least basic EF scenarios should work with Azure Synapse as the latter supports T-SQL. If you run into any specific problems, please open separate issues for these and we'll investigate.

**Comment by ErikEJ** [comment]
@ilazarevski I also did an effort a while ago to make database first scaffolding work with "Azure SQL Data Warehouse"

**Comment by ilazarevski** [comment]
@ErikEJ Nice to hear and thanks for the information. I will check and if it necessary I will post back.

**Comment by cjblomqvist** [comment]
I'm getting worried Value Objects won't make it for 7.0 despite being part of the plan. I'm sure you're well aware and haven't forgotten about it, but just a friendly reminder that it's a feature I'm (and lots of others) very much looking forward to! 😊

**Comment by ajcvickers** [comment]
@cjblomqvist Unfortunately, value objects is not going to make it. This was a painful cut to make.  We know people are waiting on this, and it will be a high priority for EF8.

**Comment by AraHaan** [comment]
Oh no, if only someone could have helped with implementing value objects so then it could have made it in.

**Comment by DerGuru** [comment]
I am very sad, that bulk inserts are not there. Any specific reason, except maybe constraints?

**Comment by ErikEJ** [comment]
@DerGuru inserts are batched in SaveChanges already. Or are you thinking of something else?

**Comment by roji** [comment]
@DerGuru @ErikEJ there's #27333 for a bulk import API over e.g. SqlBulkCopy; this hasn't been done because there have been other more important and highly-voted features. But as @ErikEJ wrote, our regular SaveChanges batches and the performance isn't too bad, unless you're really importing huge amounts of data.

**Comment by DerGuru** [comment]
> @DerGuru @ErikEJ there's #27333 unless you're really importing huge amounts of data.

I don't know... 60000 to 100000 rows of data.

**Comment by roji** [comment]
A real bulk import (via e.g. SqlBulkCopy) will likely perform noticeably better for this, though if it's a relatively one-time thing, the difference shouldn't be important and EF should be good enough. I suggest just setting up a quick benchmark yourself, it's really easy (use BenchmarkDotNet).

**Comment by DerGuru** [comment]
> @DerGuru inserts are batched in SaveChanges already. Or are you thinking of something else?

Batches are not the same, as one single SQL statement. For large quantities the difference is significant. I am talking about 50000-100000 rows, which we chunk into 1000, currently, as part of some automated task.

**Comment by roji** [comment]
@DerGuru when inserting multiple rows into the same table, EF Core does combine multiple insertions into a single INSERT or MERGE command. This is still not as efficient as using SqlBulkCopy, which is a high-speed data import mechanism, but it should be already quite optimized.

In any case, I don't think it makes sense to continue this conversation on this issue. We already have #27333 covering an abstraction over SqlBulkCopy, and apart from that the regular SaveChanges is already quite optimized (e.g. via MERGE).

**Comment by DerGuru** [comment]
You are right…

**Comment by ErikEJ** [comment]
@pinkfloydx33 .NET 6

**Comment by pinkfloydx33** [comment]
Really? Great news! Thanks

**Comment by SokoFromNZ** [comment]
When will this be available/fixed? We are struggling with the Async performance of blob data
